libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+    struct SRTPContext srtp_rtcp_send;
+    /* The SRTP receive context, to decrypt incoming packets. */
+    struct SRTPContext srtp_recv;
+
+    /* The time jitter base for audio OPUS stream. */
+    int64_t audio_jitter_base;
+
+    /* The UDP transport is used for delivering ICE, DTLS and SRTP packets. */
+    URLContext *udp_uc;
+
+    /* The maximum number of retries for ICE transmission. */
+    int ice_arq_max;
+    /* The step start timeout in ms for ICE transmission. */
+    int ice_arq_timeout;
+    /* The maximum number of retries for DTLS transmission. */
+    int dtls_arq_max;
+    /* The step start timeout in ms for DTLS transmission. */
+    int dtls_arq_timeout;
+    /**
+     * The size of RTP packet, should generally be set to MTU.
+     * Note that pion requires a smaller value, for example, 1200.
+     */
+    int pkt_size;
+    /**
+     * The optional Bearer token for WHIP Authorization.
+     * See https://www.ietf.org/archive/id/draft-ietf-wish-whip-08.html#name-authentication-and-authoriz
+     */
+    char* authorization;
+} RTCContext;
+
+static int on_rtp_write_packet(void *opaque, uint8_t *buf, int buf_size);
+
+/**
+ * Initialize and check the options for the WebRTC muxer.
+ */
+static av_cold int whip_init(AVFormatContext *s)
+{
+    int ret, ideal_pkt_size = 532;
+    RTCContext *rtc = s->priv_data;
+
+    /* Use the same logging context as AV format. */
+    rtc->dtls_ctx.log_avcl = s;
+    rtc->dtls_ctx.dtls_arq_max = rtc->dtls_arq_max;
+    rtc->dtls_ctx.dtls_arq_timeout = rtc->dtls_arq_timeout;
+    rtc->dtls_ctx.pkt_size = rtc->pkt_size;
+
+    if ((ret = dtls_context_init(&rtc->dtls_ctx)) < 0) {
+        av_log(s, AV_LOG_ERROR, "WHIP: Failed to init DTLS context\n");
+        return ret;
+    }
+
+    av_log(s, AV_LOG_INFO, "WHIP: Init ice_arq_max=%d, ice_arq_timeout=%d, dtls_arq_max=%d, dtls_arq_timeout=%d pkt_size=%d\n",
+        rtc->ice_arq_max, rtc->ice_arq_timeout, rtc->dtls_arq_max, rtc->dtls_arq_timeout, rtc->pkt_size);
+
+    if (rtc->pkt_size < ideal_pkt_size)
+        av_log(s, AV_LOG_WARNING, "WHIP: pkt_size=%d(<%d) is too small, may cause packet loss\n",
+            rtc->pkt_size, ideal_pkt_size);
+
+    return 0;
+}
+
+/**
+ * Parses the ISOM AVCC format of extradata and extracts SPS/PPS.
+ *
+ * This function is used to parse SPS/PPS from the extradata in ISOM AVCC format.
+ * It can handle both ISOM and annexb formats but only parses data in ISOM format.
+ * If the extradata is in annexb format, this function ignores it, and uses the entire
+ * extradata as a sequence header with SPS/PPS. Refer to ff_isom_write_avcc.
+ *
+ * @param s                Pointer to the AVFormatContext
+ * @param extradata        Pointer to the extradata
+ * @param extradata_size   Size of the extradata
+ * @returns                Returns 0 if successful or AVERROR_xxx in case of an error.
+ */
+static int isom_read_avcc(AVFormatContext *s, uint8_t *extradata, int  extradata_size)
```

Comment:
Kieran Kunhya:

```text
This code is duplicated from other parts of FFmpeg
```

Reply:
This function is similar to ff_isom_write_avcc, which is used to encode SPS/PPS to extradata in ISOM format. This function is used to decode SPS/PPS from extradata in ISOM format.

Maybe we should extract this function to avc.c?